app/web_ui/src/routes/(app)/settings/manage_tools/[project_id]/add_tools/remote_mcp/+page.svelte

```diff
@@ -0,0 +1,252 @@
+      description?: string
+      server_url?: string
+      headers?: HeaderPair[]
+    }
+
+    if (state.name && !name) {
+      name = state.name
+    }
+
+    if (state.description && !description) {
+      description = state.description
+    }
+
+    if (state.server_url && !server_url) {
+      server_url = state.server_url
+    }
+
+    if (state.headers && headers.length === 0) {
+      headers = [...state.headers]
+    }
+  }
+
+  function buildHeadersObject(): Record<string, string> {
+    const headersObj: Record<string, string> = {}
+
+    for (const header of headers) {
+      if (header.key.trim() && header.value.trim()) {
+        headersObj[header.key.trim()] = header.value.trim()
+      }
+    }
+
+    return headersObj
+  }
+
+  async function connect_remote_mcp() {
+    try {
+      error = null
+      submitting = true
+
+      // Validate required fields
+      if (!name.trim()) {
+        throw new Error("Name is required")
+      }
+      if (!server_url.trim()) {
+        throw new Error("Server URL is required")
+      }
+
+      // Enforce absolute http(s) URLs only
+      try {
+        const u = new URL(server_url.trim())
+        if (u.protocol !== "https:" && u.protocol !== "http:") {
+          throw new Error("Server URL must start with http:// or https://")
+        }
+      } catch {
+        throw new Error("Server URL is not a valid URL")
+      }
+
+      if (headers.length > 0) {
+        for (const header of headers) {
+          const key = header.key.trim()
+          const value = header.value.trim()
+          if (!key) {
+            throw new Error("Header name is required")
+          }
+          if (!value) {
+            throw new Error("Header value is required")
+          }
+
+          // Reject invalid header names and CR/LF in names/values
+          const tokenRe = /^[!#$%&'*+.^_`|~0-9A-Za-z-]+$/
+          if (!tokenRe.test(key)) {
+            throw new Error(`Invalid header name: "${key}"`)
+          }
+          if (/\r|\n/.test(key) || /\r|\n/.test(value)) {
+            throw new Error(
+              "Header names/values must not contain invalid characters",
+            )
+          }
+        }
+      }
+
+      const headersObj = buildHeadersObject()
+
+      //  Make a request to the server and make sure it's valid
+      try {
+        const response = await fetch(server_url.trim(), {
+          headers: headersObj,
+        })
+        if (!response.ok) {
+          throw new Error(
+            `Server returned error status: ${response.status} ${response.statusText}`,
+          )
+        }
+      } catch (e) {
+        // Log custom error message
+        const error = e as Error
+        throw new Error(
+          `${error.message}. Unable to connect to the server. Please check the URL, headers and ensure the server is accessible`,
+        )
+      }
+
+      const { data, error: api_error } = await client.POST(
+        "/api/projects/{project_id}/connect_remote_mcp",
```

Comment:
put your validation (including server check) inside this API. Make it return really helpful errors like you do here and those will bubble up the same way in the catch block below.